### PR TITLE
Various improvements

### LINKED ISSUE #2357: Dirty reactor when running Tribler.Test.test_my_channel.TestMyChannel.test_add_torrents_playlists
```
DEBUG   1467200224  test_as_server:TestMyChannel:630  Thread still running MainThread, daemon False, instance: <_MainThread(MainThread, started 139982354573056)>
DEBUG   1467200224  test_as_server:TestMyChannel:630  Thread still running Twisted, daemon True, instance: <Thread(Twisted, started daemon 139982230808320)>
DEBUG   1467200224  test_as_server:TestMyChannel:630  Thread still running WatchDog, daemon True, instance: <WatchDog(WatchDog, started daemon 139981869360896)>
DEBUG   1467200224  test_as_server:TestMyChannel:630  Thread still running StandaloneEndpoint, daemon True, instance: <Thread(StandaloneEndpoint, started daemon 139981465036544)>
DEBUG   1467200224  test_as_server:TestMyChannel:630  Thread still running PoolThread-twisted.internet.reactor-0, daemon True, instance: <Thread(PoolThread-twisted.internet.reactor-0, started daemon 139981851789056)>
DEBUG   1467200224  test_as_server:TestMyChannel:634  Content of pymdht.log
DEBUG   1467200224  test_as_server:TestMyChannel:641  Finished printing content of pymdht.log
INFO    1467200224         _legacy:twisted:154  (TCP Port 23946 Closed)
INFO    1467200224         _legacy:twisted:154  Stopping factory <twisted.web.server.Site instance at 0x7f500072fcf8>
ERROR   1467200224  test_as_server:TestMyChannel:138  The reactor was dirty:
ERROR   1467200224  test_as_server:TestMyChannel:140  >     <DelayedCall 0x7f5010e9f8c0 [225.245066166s] called=0 cancelled=0 LoopingCall<300>(DiscoveryCommunity.periodically_insert_trackers, *(), **{})()>
ERROR   1467200224  test_as_server:TestMyChannel:140  >     <DelayedCall 0x7f50002709e0 [225.249401093s] called=0 cancelled=0 LoopingCall<300>(DiscoveryCommunity.periodically_insert_trackers, *(), **{})()>
--------------------- >> end captured logging << ---------------------
```

### LINKED ISSUE #2614: Error when creating channel with long unicode description field
On [latest alpha](https://jenkins.tribler.org/job/Build_Tribler_devel/155/)...

Test data:
tribler_channels_in_7__nov2016

```
WARNING:TunnelLogger:Got ping (1403386228) from ('69.206.64.73', 7759), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
WARNING:TunnelLogger:Got pong (104852258) from ('95.211.155.142', 54025), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
Unhandled error in Deferred:
Temporarily disabling observer LegacyLogObserverWrapper(<function unhandled_error_observer at 0x7f590807d5f0>) due to exception: [Failure instance: Traceback: <type 'exceptions.KeyError'>: 'log_text'
/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py:252:inContext
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:697:__del__
/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py:258:critical
/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py:132:emit
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/logger/_observer.py:131:__call__
/usr/lib/python2.7/dist-packages/twisted/logger/_legacy.py:93:__call__
/usr/share/tribler/TriblerGUI/core_manager.py:25:unhandled_error_observer
]

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 252, in inContext
    inContext.theWork = None
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 697, in __del__
    isError=True)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py", line 258, in critical
    self.emit(LogLevel.critical, format, **kwargs)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py", line 132, in emit
    self.observer(event)
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_observer.py", line 131, in __call__
    observer(event)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_legacy.py", line 93, in __call__
    self.legacyObserver(event)
  File "/usr/share/tribler/TriblerGUI/core_manager.py", line 25, in unhandled_error_observer
    core_queue.put(event['log_text'])
exceptions.KeyError: 'log_text'


Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 393, in callback
    self._startRunCallbacks(result)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/share/tribler/Tribler/Core/Modules/restapi/channels/channels_torrents_endpoint.py", line 195, in _on_torrent_def_loaded
    self.session.add_torrent_def_to_channel(channel[0], torrent_def, extra_info, forward=True)
  File "/usr/share/tribler/Tribler/Core/Session.py", line 780, in add_torrent_def_to_channel
    forward=forward)
  File "/usr/share/tribler/Tribler/community/channel/community.py", line 363, in _disp_create_torrent
    self._dispersy.store_update_forward([message], store, update, forward)
  File "/usr/share/tribler/Tribler/dispersy/dispersy.py", line 1716, in store_update_forward
    self._store(messages)
  File "/usr/share/tribler/Tribler/dispersy/util.py", line 150, in wrapper
    return_value = func(*args, **kargs)
  File "/usr/share/tribler/Tribler/dispersy/dispersy.py", line 1552, in _store
    ), get_lastrowid=True)
  File "/usr/share/tribler/Tribler/dispersy/util.py", line 150, in wrapper
    return_value = func(*args, **kargs)
  File "/usr/share/tribler/Tribler/dispersy/database.py", line 280, in execute
    assert self._debug_thread_ident == thread.get_ident(), "Calling Database.execute on the wrong thread"
exceptions.AssertionError: Calling Database.execute on the wrong thread
Temporarily disabling observer LegacyLogObserverWrapper(<function unhandled_error_observer at 0x7f590807d5f0>) due to exception: [Failure instance: Traceback: <type 'exceptions.KeyError'>: 'log_text'
/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py:252:inContext
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:707:__del__
/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py:178:failure
/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py:132:emit
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/logger/_observer.py:131:__call__
/usr/lib/python2.7/dist-packages/twisted/logger/_legacy.py:93:__call__
/usr/share/tribler/TriblerGUI/core_manager.py:25:unhandled_error_observer
]

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 252, in inContext
    inContext.theWork = None
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 707, in __del__
    debugInfo=debugInfo)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py", line 178, in failure
    self.emit(level, format, log_failure=failure, **kwargs)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_logger.py", line 132, in emit
    self.observer(event)
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_observer.py", line 131, in __call__
    observer(event)
  File "/usr/lib/python2.7/dist-packages/twisted/logger/_legacy.py", line 93, in __call__
    self.legacyObserver(event)
  File "/usr/share/tribler/TriblerGUI/core_manager.py", line 25, in unhandled_error_observer
    core_queue.put(event['log_text'])
exceptions.KeyError: 'log_text'
WARNING:TunnelLogger:Got ping (1342292844) from ('142.165.249.197', 7759), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
WARNING:TunnelLogger:Got pong (1152620393) from ('95.211.155.142', 49630), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
WARNING:TunnelLogger:Got ping (3876804879) from ('90.206.251.145', 44356), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
WARNING:TunnelLogger:Got pong (2645985973) from ('95.211.155.142', 54025), I am <Member 6 78e7f393e3256a792f96ca5a841cb0115740b5ba>
WARNING:SearchCommunity:message from {10.0.0.1:7759 194.96.20.128:7759} cannot be found in the request cache, skipping it
WARNING:Timeline:FAIL time:6 user:7836 -> permit^channel (no authorization)
WARNING:Timeline:FAIL time:5 user:7837 -> permit^channel (no authorization)

```

### LINKED ISSUE #2615: exceptions.UnicodeDecodeError: 'utf8' codec can't decode byte 0xa1 in position 34: invalid start byte
While running latest alpha, after Dispersy already gave another TraceBack and perhaps died.

```
WARNING:Timeline:FAIL time:6 user:17610 -> permit^channel (no authorization)
WARNING:Timeline:FAIL time:6 user:17611 -> permit^channel (no authorization)
WARNING:Timeline:FAIL time:6 user:17612 -> permit^channel (no authorization)
WARNING:SearchCommunity:message from {192.168.254.16:7759 14.2.77.217:7759} cannot be found in the request cache, skipping it
WARNING:Timeline:FAIL time:6 user:17615 -> permit^channel (no authorization)
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/dist-packages/twisted/_threads/_threadworker.py", line 46, in work
    task()
  File "/usr/lib/python2.7/dist-packages/twisted/_threads/_team.py", line 190, in doWork
    task()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 246, in inContext
    result = inContext.theWork()
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 262, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/share/tribler/Tribler/Core/Modules/restapi/events_endpoint.py", line 139, in on_torrent_discovered
    self.write_data(json.dumps({"type": "torrent_discovered", "event": args[0]}))
  File "/usr/lib/python2.7/json/__init__.py", line 244, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
exceptions.UnicodeDecodeError: 'utf8' codec can't decode byte 0xa1 in position 34: invalid start byte

WARNING:AllChannelCommunity:drop a 123 byte packet duplicate message by member^global_time (1) [dispersy-undo-own] from {192.168.0.3:7759 50.89.188.120:7759}
WARNING:AllChannelCommunity:drop a 123 byte packet duplicate message by member^global_time (1) [dispersy-undo-own] from {192.168.0.3:7759 50.89.188.120:7759}
ERROR:root:Traceback (most recent call last):
  File "/usr/share/tribler/TriblerGUI/core_manager.py", line 101, in check_queue
    raise RuntimeError(data)
RuntimeError: Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/dist-packages/twisted/_threads/_threadworker.py", line 46, in work
    task()
  File "/usr/lib/python2.7/dist-packages/twisted/_threads/_team.py", line 190, in doWork
    task()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 246, in inContext
    result = inContext.theWork()
  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 262, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/share/tribler/Tribler/Core/Modules/restapi/events_endpoint.py", line 139, in on_torrent_discovered
    self.write_data(json.dumps({"type": "torrent_discovered", "event": args[0]}))
  File "/usr/lib/python2.7/json/__init__.py", line 244, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
exceptions.UnicodeDecodeError: 'utf8' codec can't decode byte 0xa1 in position 34: invalid start byte


WARNING:TunnelLogger:Removing relay 3850354308 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 927781133 unload
WARNING:TunnelLogger:Removing relay 2826112533 unload
WARNING:TunnelLogger:Removing relay 1188879011 unload
WARNING:TunnelLogger:Removing relay 1522936104 unload
WARNING:TunnelLogger:Removing relay 3433527609 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 153103290 unload
WARNING:TunnelLogger:Removing relay 4205191621 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 4195962125 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 1353309523 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 794590036 unload
WARNING:TunnelLogger:Removing relay 1199687260 unload
WARNING:TunnelLogger:Removing relay 3888567266 unload
WARNING:TunnelLogger:Removing relay 2765509479 unload
WARNING:TunnelLogger:Removing relay 1136348264 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 2946991986 unload
WARNING:TunnelLogger:Removing relay 967759989 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Removing relay 890742783 unload
WARNING:TunnelLogger:MULTICHAIN: Tunnel candidate not found
WARNING:TunnelLogger:Received message for unknown circuit ID: 1353309523
WARNING:TunnelLogger:Received message for unknown circuit ID: 2946991986
WARNING:TunnelLogger:Received message for unknown circuit ID: 2946991986
ERROR:root:Traceback (most recent call last):
  File "/usr/share/tribler/TriblerGUI/widgets/editchannelpage.py", line 234, in on_torrent_to_channel_added
    if 'added' in result:
TypeError: argument of type 'NoneType' is not iterable

Traceback (most recent call last):
  File "/usr/share/tribler/Tribler/Core/APIImplementation/LaunchManyCore.py", line 785, in early_shutdown
ERROR:root:No JSON object could be decoded
Traceback (most recent call last):
  File "/usr/share/tribler/TriblerGUI/tribler_request_manager.py", line 74, in on_finished
    json_result = json.loads(str(data))
  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
    success = yield self.dispersy.stop()
RuntimeError: Dispersy is not running
Stopping Tribler...
Tribler stopped!!
Stopping Tribler...
```

### PR DESCRIPTION
**Pull request by devos50**
Special thanks to @brussee for the helpful suggestions!

This PR contains the following:
-Fixes #2614 
-Removed events buffer
-Fixed unicode issues when sending responses over the API (Fixes #2615)
-Fixed a bug when switching from VoD files
-Added caching to the torrent info endpoint (when fetching the same magnet twice, it will load the one from cache)
-Written tests for the tracker manager
-Completely refactored the torrent checker so it makes better use of Twisted
-Created a global exception handler that sends exceptions over the HTTP API when they are catched
-Written a state endpoint that returns the current state of the Tribler core and whether an exception has occurred
-Written an endpoint to fetch the health of a swarm
-Implemented a HTTP/UDP tracker that can be used in unit tests
-Fixes #2357

### CODE REVIEW

File: Tribler/Test/Core/Modules/RestApi/test_state_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+class TestStateEndpoint(AbstractApiTest):
+
+    @deferred(timeout=10)
+    def test_get_state(self):
+        """
+        Testing whether the API returns a correct state when requested
+        """
+        self.session.lm.api_manager.root_endpoint.state_endpoint.on_tribler_exception("abcd")
+        expected_json = {"state": "STARTED", "last_exception": "abcd"}
```
Comment (brussee) [inline Tribler/Test/Core/Modules/RestApi/test_state_endpoint.py]: Should this not say: ``{"state": "EXCEPTION",``

File: Tribler/Core/Modules/restapi/state_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            .. sourcecode:: none
+
+                curl -X GET http://localhost:8085/state
+
+            **Example response**:
+
+            .. sourcecode:: javascript
+
+                {
+                    "state": "STARTED"
```
Comment (brussee) [inline Tribler/Core/Modules/restapi/state_endpoint.py]: +``"last_exception": None``

File: Tribler/Test/Core/Modules/RestApi/test_state_endpoint.py (hunk 1 above)
Comment (devos50) [inline Tribler/Test/Core/Modules/RestApi/test_state_endpoint.py]: Yes it should, thanks for noticing!

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Pull request merged**